Userland/Libraries/LibGfx/ImageFormats/ExifOrientedBitmap.h

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2023, Lucas Chollet <lucas.chollet@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/NonnullOwnPtr.h>
+#include <LibGfx/Bitmap.h>
+
+namespace Gfx {
+class ExifOrientedBitmap {
+public:
+    // In the EXIF 3.0 specification, 4.6.5.1.6. Orientation
+    enum class Orientation {
+        None = 1,
+        FlipHorizontally = 2,
+        Rotate180 = 3,
+        FlipVertically = 4,
+        Rotate90ClockwiseThenFlipHorizontally = 5,
+        Rotate90Clockwise = 6,
+        FlipHorizontallyThenRotate90Clockwise = 7,
+        Rotate90CounterClockwise = 8,
+    };
+
+    template<typename... Args>
+    static ErrorOr<ExifOrientedBitmap> create(BitmapFormat format, IntSize size, Orientation orientation)
+    {
+        auto bitmap = TRY(Bitmap::create(format, oriented_size(size, orientation)));
+        return ExifOrientedBitmap(move(bitmap), size, orientation);
+    }
+
+    void set_pixel(u32 x, u32 y, Color color)
+    {
+        auto const new_position = oriented_position(IntPoint(x, y));
+        m_bitmap->set_pixel(new_position, color);
```

Comment:
This looks like a very slow approach.

How about decoding into a normal bitmap, and then transforming the bitmap in one go after it has been decided? Then you have the orientation check once, instead of once per pixel.

Reply:
Indeed, it's not designed with peak efficiency in mind.
I'm pretty sure it's not the main bottleneck right now and even if it is, I don't think that it should be a blocker.
Your proposition can be implemented within the class with very little change to the API, and anyway if we need to squeeze performance out of it, I would rather try a template parameter.

Reply:
Why not go with the normal approach though? Having just the enum and a `RefPtr reorient_bitmap(Bitmap, Orientation)` function is likely less code, faster, _and_ easier to understand (as it detangles decoding and reorientation).

(Also makes it easier to move to paint-time reorientation, should we want to do that in the future.)